# 1. Import packages

In [ ]:
import nltk
nltk.download()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------


Downloader>  d



Download which package (l=list; x=cancel)?


  Identifier>  l


Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [ ] cess_esp............ CESS-ESP Treebank
  [ ] chat80.............. Chat-80 Data Files
  [ ] city_database....... City Database
  [ ] cmudict............. The Carnegie Mellon Pronouncing Dictionary (0.6)
  [ ] comparative_sentences Comparative Sentence Dataset
  [ ] comtrans............ ComTrans Corpus Sample
 

Hit Enter to continue:  


  [ ] conll2007........... Dependency Treebanks from CoNLL 2007 (Catalan
                           and Basque Subset)
  [ ] crubadan............ Crubadan Corpus
  [ ] dependency_treebank. Dependency Parsed Treebank
  [ ] dolch............... Dolch Word List
  [ ] europarl_raw........ Sample European Parliament Proceedings Parallel
                           Corpus
  [ ] floresta............ Portuguese Treebank
  [ ] framenet_v15........ FrameNet 1.5
  [ ] framenet_v17........ FrameNet 1.7
  [ ] gazetteers.......... Gazeteer Lists
  [ ] genesis............. Genesis Corpus
  [ ] gutenberg........... Project Gutenberg Selections
  [ ] ieer................ NIST IE-ER DATA SAMPLE
  [ ] inaugural........... C-Span Inaugural Address Corpus
  [ ] indian.............. Indian Language POS-Tagged Corpus
  [ ] jeita............... JEITA Public Morphologically Tagged Corpus (in
                           ChaSen format)
  [ ] kimmo............... PC-KIMMO Data Files
  [ ] knbc................ KNB C

Hit Enter to continue:  


  [ ] lin_thesaurus....... Lin's Dependency Thesaurus
  [ ] mac_morpho.......... MAC-MORPHO: Brazilian Portuguese news text with
                           part-of-speech tags
  [ ] machado............. Machado de Assis -- Obra Completa
  [ ] masc_tagged......... MASC Tagged Corpus
  [ ] maxent_ne_chunker... ACE Named Entity Chunker (Maximum entropy)
  [ ] maxent_treebank_pos_tagger Treebank Part of Speech Tagger (Maximum entropy)
  [ ] moses_sample........ Moses Sample Models
  [ ] movie_reviews....... Sentiment Polarity Dataset Version 2.0
  [ ] mte_teip5........... MULTEXT-East 1984 annotated corpus 4.0
  [ ] mwa_ppdb............ The monolingual word aligner (Sultan et al.
                           2015) subset of the Paraphrase Database.
  [ ] names............... Names Corpus, Version 1.3 (1994-03-29)
  [ ] nombank.1.0......... NomBank Corpus 1.0
  [ ] nonbreaking_prefixes Non-Breaking Prefixes (Moses Decoder)
  [ ] nps_chat............ NPS Chat
  [ ] omw................. Open Mu

Hit Enter to continue:  


  [ ] perluniprops........ perluniprops: Index of Unicode Version 7.0.0
                           character properties in Perl
  [ ] pil................. The Patient Information Leaflet (PIL) Corpus
  [ ] pl196x.............. Polish language of the XX century sixties
  [ ] porter_test......... Porter Stemmer Test Files
  [ ] ppattach............ Prepositional Phrase Attachment Corpus
  [ ] problem_reports..... Problem Report Corpus
  [ ] product_reviews_1... Product Reviews (5 Products)
  [ ] product_reviews_2... Product Reviews (9 Products)
  [ ] propbank............ Proposition Bank Corpus 1.0
  [ ] pros_cons........... Pros and Cons
  [ ] ptb................. Penn Treebank
  [ ] punkt............... Punkt Tokenizer Models
  [ ] qc.................. Experimental Data for Question Classification
  [ ] reuters............. The Reuters-21578 benchmark corpus, ApteMod
                           version
  [ ] rslp................ RSLP Stemmer (Removedor de Sufixos da Lingua
             

Hit Enter to continue:  


  [ ] senseval............ SENSEVAL 2 Corpus: Sense Tagged Text
  [ ] sentence_polarity... Sentence Polarity Dataset v1.0
  [ ] sentiwordnet........ SentiWordNet
  [ ] shakespeare......... Shakespeare XML Corpus Sample
  [ ] sinica_treebank..... Sinica Treebank Corpus Sample
  [ ] smultron............ SMULTRON Corpus Sample
  [ ] snowball_data....... Snowball Data
  [ ] spanish_grammars.... Grammars for Spanish
  [ ] state_union......... C-Span State of the Union Address Corpus
  [ ] stopwords........... Stopwords Corpus
  [ ] subjectivity........ Subjectivity Dataset v1.0
  [ ] swadesh............. Swadesh Wordlists
  [ ] switchboard......... Switchboard Corpus Sample
  [ ] tagsets............. Help on Tagsets
  [ ] timit............... TIMIT Corpus Sample
  [ ] toolbox............. Toolbox Sample Files
  [ ] treebank............ Penn Treebank Sample
  [ ] twitter_samples..... Twitter Samples
  [ ] udhr2............... Universal Declaration of Human Rights Corpus
                     

Hit Enter to continue:  stopwords


  [ ] unicode_samples..... Unicode Samples
  [ ] universal_tagset.... Mappings to the Universal Part-of-Speech Tagset
  [ ] universal_treebanks_v20 Universal Treebanks Version 2.0
  [ ] vader_lexicon....... VADER Sentiment Lexicon
  [ ] verbnet3............ VerbNet Lexicon, Version 3.3
  [ ] verbnet............. VerbNet Lexicon, Version 2.1
  [ ] webtext............. Web Text Corpus
  [ ] wmt15_eval.......... Evaluation data from WMT15
  [ ] word2vec_sample..... Word2Vec Sample
  [ ] wordnet............. WordNet
  [ ] wordnet_ic.......... WordNet-InfoContent
  [ ] words............... Word Lists
  [ ] ycoe................ York-Toronto-Helsinki Parsed Corpus of Old
                           English Prose

Collections:
  [ ] all-corpora......... All the corpora
  [ ] all-nltk............ All packages available on nltk_data gh-pages
                           branch
  [ ] all................. All packages
  [ ] book................ Everything used in the NLTK Book
  [ ] popular..........

Hit Enter to continue:  all


  [ ] tests............... Packages for running tests
  [ ] third-party......... Third-party data packages

([*] marks installed packages)

Download which package (l=list; x=cancel)?


  Identifier>  



---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------


Downloader>  d



Download which package (l=list; x=cancel)?


  Identifier>  l


Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [ ] cess_esp............ CESS-ESP Treebank
  [ ] chat80.............. Chat-80 Data Files
  [ ] city_database....... City Database
  [ ] cmudict............. The Carnegie Mellon Pronouncing Dictionary (0.6)
  [ ] comparative_sentences Comparative Sentence Dataset
  [ ] comtrans............ ComTrans Corpus Sample
 

Hit Enter to continue:  stopwords


  [ ] conll2007........... Dependency Treebanks from CoNLL 2007 (Catalan
                           and Basque Subset)
  [ ] crubadan............ Crubadan Corpus
  [ ] dependency_treebank. Dependency Parsed Treebank
  [ ] dolch............... Dolch Word List
  [ ] europarl_raw........ Sample European Parliament Proceedings Parallel
                           Corpus
  [ ] floresta............ Portuguese Treebank
  [ ] framenet_v15........ FrameNet 1.5
  [ ] framenet_v17........ FrameNet 1.7
  [ ] gazetteers.......... Gazeteer Lists
  [ ] genesis............. Genesis Corpus
  [ ] gutenberg........... Project Gutenberg Selections
  [ ] ieer................ NIST IE-ER DATA SAMPLE
  [ ] inaugural........... C-Span Inaugural Address Corpus
  [ ] indian.............. Indian Language POS-Tagged Corpus
  [ ] jeita............... JEITA Public Morphologically Tagged Corpus (in
                           ChaSen format)
  [ ] kimmo............... PC-KIMMO Data Files
  [ ] knbc................ KNB C

Hit Enter to continue:  


  [ ] lin_thesaurus....... Lin's Dependency Thesaurus
  [ ] mac_morpho.......... MAC-MORPHO: Brazilian Portuguese news text with
                           part-of-speech tags
  [ ] machado............. Machado de Assis -- Obra Completa
  [ ] masc_tagged......... MASC Tagged Corpus
  [ ] maxent_ne_chunker... ACE Named Entity Chunker (Maximum entropy)
  [ ] maxent_treebank_pos_tagger Treebank Part of Speech Tagger (Maximum entropy)
  [ ] moses_sample........ Moses Sample Models
  [ ] movie_reviews....... Sentiment Polarity Dataset Version 2.0
  [ ] mte_teip5........... MULTEXT-East 1984 annotated corpus 4.0
  [ ] mwa_ppdb............ The monolingual word aligner (Sultan et al.
                           2015) subset of the Paraphrase Database.
  [ ] names............... Names Corpus, Version 1.3 (1994-03-29)
  [ ] nombank.1.0......... NomBank Corpus 1.0
  [ ] nonbreaking_prefixes Non-Breaking Prefixes (Moses Decoder)
  [ ] nps_chat............ NPS Chat
  [ ] omw................. Open Mu

Hit Enter to continue:  


  [ ] perluniprops........ perluniprops: Index of Unicode Version 7.0.0
                           character properties in Perl
  [ ] pil................. The Patient Information Leaflet (PIL) Corpus
  [ ] pl196x.............. Polish language of the XX century sixties
  [ ] porter_test......... Porter Stemmer Test Files
  [ ] ppattach............ Prepositional Phrase Attachment Corpus
  [ ] problem_reports..... Problem Report Corpus
  [ ] product_reviews_1... Product Reviews (5 Products)
  [ ] product_reviews_2... Product Reviews (9 Products)
  [ ] propbank............ Proposition Bank Corpus 1.0
  [ ] pros_cons........... Pros and Cons
  [ ] ptb................. Penn Treebank
  [ ] punkt............... Punkt Tokenizer Models
  [ ] qc.................. Experimental Data for Question Classification
  [ ] reuters............. The Reuters-21578 benchmark corpus, ApteMod
                           version
  [ ] rslp................ RSLP Stemmer (Removedor de Sufixos da Lingua
             

Hit Enter to continue:  


  [ ] senseval............ SENSEVAL 2 Corpus: Sense Tagged Text
  [ ] sentence_polarity... Sentence Polarity Dataset v1.0
  [ ] sentiwordnet........ SentiWordNet
  [ ] shakespeare......... Shakespeare XML Corpus Sample
  [ ] sinica_treebank..... Sinica Treebank Corpus Sample
  [ ] smultron............ SMULTRON Corpus Sample
  [ ] snowball_data....... Snowball Data
  [ ] spanish_grammars.... Grammars for Spanish
  [ ] state_union......... C-Span State of the Union Address Corpus
  [ ] stopwords........... Stopwords Corpus
  [ ] subjectivity........ Subjectivity Dataset v1.0
  [ ] swadesh............. Swadesh Wordlists
  [ ] switchboard......... Switchboard Corpus Sample
  [ ] tagsets............. Help on Tagsets
  [ ] timit............... TIMIT Corpus Sample
  [ ] toolbox............. Toolbox Sample Files
  [ ] treebank............ Penn Treebank Sample
  [ ] twitter_samples..... Twitter Samples
  [ ] udhr2............... Universal Declaration of Human Rights Corpus
                     

Hit Enter to continue:  stopwords


  [ ] unicode_samples..... Unicode Samples
  [ ] universal_tagset.... Mappings to the Universal Part-of-Speech Tagset
  [ ] universal_treebanks_v20 Universal Treebanks Version 2.0
  [ ] vader_lexicon....... VADER Sentiment Lexicon
  [ ] verbnet3............ VerbNet Lexicon, Version 3.3
  [ ] verbnet............. VerbNet Lexicon, Version 2.1
  [ ] webtext............. Web Text Corpus
  [ ] wmt15_eval.......... Evaluation data from WMT15
  [ ] word2vec_sample..... Word2Vec Sample
  [ ] wordnet............. WordNet
  [ ] wordnet_ic.......... WordNet-InfoContent
  [ ] words............... Word Lists
  [ ] ycoe................ York-Toronto-Helsinki Parsed Corpus of Old
                           English Prose

Collections:
  [ ] all-corpora......... All the corpora
  [ ] all-nltk............ All packages available on nltk_data gh-pages
                           branch
  [ ] all................. All packages
  [ ] book................ Everything used in the NLTK Book
  [ ] popular..........

Hit Enter to continue:  all


  [ ] tests............... Packages for running tests
  [ ] third-party......... Third-party data packages

([*] marks installed packages)

Download which package (l=list; x=cancel)?


  Identifier>  tests


       | 
       | Downloading package averaged_perceptron_tagger to
       |     /home/jupyter/nltk_data...
       |   Unzipping taggers/averaged_perceptron_tagger.zip.
       | Downloading package porter_test to /home/jupyter/nltk_data...
       |   Unzipping stemmers/porter_test.zip.
       | Downloading package twitter_samples to
       |     /home/jupyter/nltk_data...
       |   Unzipping corpora/twitter_samples.zip.
       | Downloading package wmt15_eval to /home/jupyter/nltk_data...
       |   Unzipping models/wmt15_eval.zip.
       | Downloading package subjectivity to
       |     /home/jupyter/nltk_data...
       |   Unzipping corpora/subjectivity.zip.
       | Downloading package framenet_v17 to
       |     /home/jupyter/nltk_data...
       |   Unzipping corpora/framenet_v17.zip.
       | Downloading package product_reviews_1 to
       |     /home/jupyter/nltk_data...
       |   Unzipping corpora/product_reviews_1.zip.
       | Downloading package product_reviews_2 to
    


---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------


In [1]:
import time
import csv
import re
import numpy as np
import pandas as pd
from pprint import pprint

# Gensim
import gensim
import gensim.corpora as corpora
from gensim.utils import simple_preprocess
from gensim.models import CoherenceModel

# spacy for lemmatization
import spacy

# Plotting tools
import pyLDAvis
import pyLDAvis.gensim  # don't skip this
import matplotlib.pyplot as plt
# %matplotlib inline

# Enable logging for gensim - optional
import logging
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.ERROR)

import warnings
warnings.filterwarnings("ignore",category=DeprecationWarning)

In [35]:
# NLTK Stop words
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/jupyter/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [2]:
from nltk.corpus import stopwords
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])

# 2.  Import data - monthly divided merged comments

In [3]:
# Import Dataset
month = '2019_02'
file_path = 'mergedComments_{}.json'.format(month)
df = pd.read_json(file_path,lines=True)
print(df.author)
df.head()

0                   ------sb
1         -----iMartijn-----
2            -----truth-----
3                 ----__----
4               ----lost----
                 ...        
199477            Rec0nSl0th
199478                Rec0nd
199479        Rec_desk_phone
199480    RecalcitrantAussie
199481           ReccyNegika
Name: author, Length: 199482, dtype: object


author                                         mergedText
0            ------sb  People will downvote as they see fit, but post...
1  -----iMartijn-----  Aren’t you disappointed it’s only sixteen?...
2     -----truth-----  &gt;wapo\n\nHAHAHAH,Dead people and illegal im...
3          ----__----                                             Maybe?
4        ----lost----  &gt; Could be a potentially Presidential endin...

# 3. Preprocessing
Let’s tokenize each sentence into a list of words, removing punctuations and unnecessary characters altogether.
Gensim’s **simple_preprocess()** is great for this. Additionally I have set **deacc=True** to remove the punctuations.

Here the **turncated_len** is used to indicate the length of mergedText that to be converted to list, because the **.tolist()** processing to the whole corpus will be a little time-comsuming, i.e. about several minutes. However, the whole corpus should keep intact because it will influence the following steps, of course you can save the time if you just try this model for fun.

## Cell duration 166s

In [4]:
# Convert to list
start = time.time()

# turncated_len = 2000
data = df.mergedText.values.tolist()

# Remove new line characters. Here '\s' matches all the blank characters
data = [re.sub('\s+', ' ', item) for item in data]

# Remove distracting single quotes
data = [re.sub("\'", "", item) for item in data]
pprint(data[:1])

# Tokenize words and clean-up text
def item_to_words(sentences):
    for sentence in sentences:
        yield(gensim.utils.simple_preprocess(str(sentence), deacc=True))  # deacc=True removes punctuations

print('Item to words')
data_words = list(item_to_words(data))
print(data_words[:1])

end = time.time()
print ('Cell duration')
print(end - start)

['People will downvote as they see fit, but posting a comment that is nothing '
 'but yet another complaint about downvotes - that’s a paddlin’. ,That’s '
 'a bribe and conservative governments the world over have it down to an art. '
 'Throw a small bone that will help in the short term to secure votes, and '
 'propagate the misconception that they manage the economy better, and then '
 'build mechanisms to systematically syphon money to the wealthy for decades '
 'to come. ']
Item to words
[['people', 'will', 'downvote', 'as', 'they', 'see', 'fit', 'but', 'posting', 'comment', 'that', 'is', 'nothing', 'but', 'yet', 'another', 'complaint', 'about', 'downvotes', 'thata', 'paddlina', 'thata', 'bribe', 'and', 'conservative', 'governments', 'the', 'world', 'over', 'have', 'it', 'down', 'to', 'an', 'art', 'throw', 'small', 'bone', 'that', 'will', 'help', 'in', 'the', 'short', 'term', 'to', 'secure', 'votes', 'and', 'propagate', 'the', 'misconception', 'that', 'they', 'manage', 'the',

## The number of authors and the number of the total words

In [5]:
print('The number of authors')
print(len(data_words))
count_original = 0
for data_word in data_words:
    count_original += len(data_word)
print('The number ot the total words')
count_original

The number of authors
199482
The number ot the total words


66115290

# 4. Creating Bigram and Trigram Models
## Note the variable name for the dataset up till now
**df** is the original imported data which we have not changed it   
**data** is an intermediate variable which have been slightly preprocessed    
**data_words** is the thing to be used in the following which have been completely cleaned and converted to a list of words   
## This part is time-consuming as the cells durations present

### Cell duration 115s

In [11]:
# Build the bigram model
start = time.time()
bigram = gensim.models.Phrases(data_words, min_count=5, threshold=100) # higher threshold fewer phrases.
end = time.time()
print ('Duration of the bigram building')
print(end - start)

Duration of the bigram building
102.34177756309509


### Cell duration 422s

In [6]:
# Build the trigram model
start = time.time()
trigram = gensim.models.Phrases(bigram[data_words], threshold=100) 
end = time.time()
print ('Duration of the trigram building')
print(end - start)

Duration of the trigram building
401.48244047164917


### Cell duration 77s

In [16]:
# Faster way to get a sentence clubbed as a trigram/bigram
start = time.time()
bigram_mod = gensim.models.phrases.Phraser(bigram)
end = time.time()
print ('Duration of the trigram building')
print(end - start)

Duration of the trigram building
62.84939098358154


### Cell duration 84s

In [8]:
# Faster way to get a sentence clubbed as a trigram/bigram
start = time.time()
trigram_mod = gensim.models.phrases.Phraser(trigram)
end = time.time()
print ('Duration of the trigram building')
print(end - start)

Duration of the trigram building
80.29017472267151


In [9]:
# See bigram example and trigram example
print(trigram_mod[bigram_mod[data_words[2]]])
print(data_words[2])

['gt', 'wapo', 'hahahah', 'dead', 'people', 'and', 'illegal_immigrants', 'dona', 'count', 'democracies', 'are', 'to', 'protect', 'against', 'tyrants', 'not', 'to', 'make', 'sure', 'every', 'sudden', 'and', 'transient', 'whim', 'of', 'the', 'majority', 'becomes', 'law', 'the', 'people', 'voting', 'for', 'donald', 'trump', 'is', 'what', 'makes', 'him', 'qualified', 'tbh', 'that', 'would', 'be', 'fucking', 'hilarious', 'going', 'to', 'tell', 'my', 'jewish', 'gym', 'buddy', 'that', 'tomorrow', 'evening', 'sure', 'hea', 'll', 'chuckle', 'the', 'usa', 'is', 'not', 'democracy', 'and', 'was', 'never', 'designed', 'or', 'intended', 'to', 'be', 'one', 'only', 'because', 'the', 'left', 'is', 'so', 'violent', 'lol', 'you', 'honestly', 'need', 'source', 'for', 'why', 'dead', 'people', 'and', 'illegal_immigrants', 'dona', 'count']
['gt', 'wapo', 'hahahah', 'dead', 'people', 'and', 'illegal', 'immigrants', 'dona', 'count', 'democracies', 'are', 'to', 'protect', 'against', 'tyrants', 'not', 'to', 'mak

# 5. Remove stopwords, make Bigrams, Lemmatize

In [13]:
# Define functions for stopwords, bigrams, trigrams and lemmatization
def remove_stopwords(texts):
    return [[word for word in simple_preprocess(str(doc)) if word not in stop_words] for doc in texts]

def make_bigrams(texts):
    return [bigram_mod[doc] for doc in texts]

def make_trigrams(texts):
    return [trigram_mod[bigram_mod[doc]] for doc in texts]

def lemmatization(texts, allowed_postags=['NOUN', 'ADJ', 'VERB', 'ADV']):
    """https://spacy.io/api/annotation"""
    texts_out = []
    start_index = 0
    process_len = 1000
    k = len(texts)//process_len
    # the process_len and range k are set to prevent the length of processing texts exceeds the spacy limit 1000000
    # and the process_len and the number of chunk k is an rough estimation, adjust it if necessary
    for num in range(k):
        print("{} / {}".format(num,k))
        start_index = num * process_len
        text_chunk = texts[start_index:(start_index+process_len)]
        for text in text_chunk:
            doc = nlp(" ".join(text)) 
            texts_out.append([token.lemma_ for token in doc if token.pos_ in allowed_postags])
    # becuase the k is round down, we need to process the last chunk
    text_last_chunk = texts[start_index+process_len:]
    for text in text_last_chunk:
            doc = nlp(" ".join(text)) 
            texts_out.append([token.lemma_ for token in doc if token.pos_ in allowed_postags])
    return texts_out

### Cell duration 191s

In [8]:
# Remove Stop Words
start = time.time()
data_words_nostops = remove_stopwords(data_words)
end = time.time()
print ('Duration of the trigram building')
print(end - start)

Duration of the trigram building
179.37500834465027


In [9]:
count_nostops = 0
for data_word in data_words_nostops:
    count_nostops += len(data_word)
print('The number of the total words without stop_words')
count_nostops

The number of the total words without stop_words


37101073

### The original corpus contains 66115290 words, by deleting the stop_words, the scale decrease half nearly.

### Cell duration 94s

In [17]:
# Form Bigrams
start = time.time()
data_words_bigrams = make_bigrams(data_words_nostops)
end = time.time()
print ('Duration of making bigrams')
print(end - start)

Duration of making bigrams
95.84422850608826


In [18]:
count_words_bigrams = 0
for data_word in data_words_bigrams:
    count_words_bigrams += len(data_word)
print('The number of the total words without bigramss')
count_words_bigrams

The number of the total words without bigramss


35705431

### After making bigrams, the sclae decreases 4%

### Cell duration 1763s

In [19]:
# Initialize spacy 'en' model, keeping only tagger component (for efficiency)
# python3 -m spacy download en
nlp = spacy.load('en', disable=['parser', 'ner'])
nlp.max_length = count_words_bigrams

In [20]:
start = time.time()
# Do lemmatization keeping only noun, adj, vb, adv
data_lemmatized = lemmatization(data_words_bigrams, allowed_postags=['NOUN', 'ADJ', 'VERB', 'ADV'])

end = time.time()
print ('Duration of Lemmatization')
print(end - start)
print(data_lemmatized[:1])

0 / 199
1 / 199
2 / 199
3 / 199
4 / 199
5 / 199
6 / 199
7 / 199
8 / 199
9 / 199
10 / 199
11 / 199
12 / 199
13 / 199
14 / 199
15 / 199
16 / 199
17 / 199
18 / 199
19 / 199
20 / 199
21 / 199
22 / 199
23 / 199
24 / 199
25 / 199
26 / 199
27 / 199
28 / 199
29 / 199
30 / 199
31 / 199
32 / 199
33 / 199
34 / 199
35 / 199
36 / 199
37 / 199
38 / 199
39 / 199
40 / 199
41 / 199
42 / 199
43 / 199
44 / 199
45 / 199
46 / 199
47 / 199
48 / 199
49 / 199
50 / 199
51 / 199
52 / 199
53 / 199
54 / 199
55 / 199
56 / 199
57 / 199
58 / 199
59 / 199
60 / 199
61 / 199
62 / 199
63 / 199
64 / 199
65 / 199
66 / 199
67 / 199
68 / 199
69 / 199
70 / 199
71 / 199
72 / 199
73 / 199
74 / 199
75 / 199
76 / 199
77 / 199
78 / 199
79 / 199
80 / 199
81 / 199
82 / 199
83 / 199
84 / 199
85 / 199
86 / 199
87 / 199
88 / 199
89 / 199
90 / 199
91 / 199
92 / 199
93 / 199
94 / 199
95 / 199
96 / 199
97 / 199
98 / 199
99 / 199
100 / 199
101 / 199
102 / 199
103 / 199
104 / 199
105 / 199
106 / 199
107 / 199
108 / 199
109 / 199
110 / 199


In [21]:
print(data_words_bigrams[:1])

print(data_lemmatized[:1])

count_words_lemma = 0
for data_word in data_lemmatized:
    count_words_lemma += len(data_word)
print('The number of the total lemmalized words')
count_words_lemma

[['people', 'downvote', 'see', 'fit', 'posting', 'comment', 'nothing', 'yet', 'another', 'complaint', 'downvotes', 'thata', 'paddlina', 'thata', 'bribe', 'conservative', 'governments', 'world', 'art', 'throw', 'small', 'bone', 'help', 'short', 'term', 'secure', 'votes', 'propagate', 'misconception', 'manage', 'economy', 'better', 'build', 'mechanisms', 'systematically', 'syphon', 'money', 'wealthy', 'decades', 'come']]
[['people', 'downvote', 'see', 'fit', 'posting', 'comment', 'yet', 'complaint', 'downvote', 'bribe', 'conservative', 'government', 'world', 'art', 'throw', 'small', 'bone', 'help', 'short', 'term', 'secure', 'vote', 'propagate', 'misconception', 'manage', 'economy', 'better', 'build', 'mechanism', 'systematically', 'syphon', 'money', 'wealthy', 'decade', 'come']]
The number of the total lemmalized words


27797154

### Lemmalized words 27797154

# 6. Create the Dictionary and Corpus needed for Topic Modeling

### Cell duration 88s

In [22]:
# Create Dictionary
start = time.time()
id2word = corpora.Dictionary(data_lemmatized)

# Create Corpus
texts = data_lemmatized

# Term Document Frequency
corpus = [id2word.doc2bow(text) for text in texts]


end = time.time()
print ('Duration of creating the dictionary and corpus')
print(end - start)

# View
print(corpus[:1])

Duration of creating the dictionary and corpus
92.77045607566833
[[(0, 1), (1, 1), (2, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1), (8, 1), (9, 1), (10, 2), (11, 1), (12, 1), (13, 1), (14, 1), (15, 1), (16, 1), (17, 1), (18, 1), (19, 1), (20, 1), (21, 1), (22, 1), (23, 1), (24, 1), (25, 1), (26, 1), (27, 1), (28, 1), (29, 1), (30, 1), (31, 1), (32, 1), (33, 1)]]


## Save id2word

In [26]:
import pickle
file_name = 'id2word_2019_02.pkl'
with open(file_name,'wb') as file_object:
    pickle.dump(id2word,file_object)

## Load id2word

In [30]:
import pickle
file_name = 'id2word_2019_02.pkl'
with open(file_name,'rb') as file_object:
    foo = pickle.load(file_object)

## Save corpus

In [33]:
import json
file_name = '_corpus_2019_02.json'
with open(file_name,'w') as file_object:
    json.dump(corpus,file_object)

## Load corpus

In [34]:
import json
file_name = '_corpus_2019_02.json'
with open(file_name,'r') as file_object:
    corpus_foo = json.load(file_object)

In [35]:
len(corpus_foo)

199482

## Test loading id2word and corpus

In [36]:
# Create Dictionary
start = time.time()
#id2word = corpora.Dictionary(data_lemmatized)

# Create Corpus
texts = data_lemmatized

# Term Document Frequency
foo_corpus = [foo.doc2bow(text) for text in texts]


end = time.time()
print ('Duration of creating the dictionary and corpus')
print(end - start)

# View
print(foo_corpus[:1])

Duration of creating the dictionary and corpus
14.78010082244873
[[(0, 1), (1, 1), (2, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1), (8, 1), (9, 1), (10, 2), (11, 1), (12, 1), (13, 1), (14, 1), (15, 1), (16, 1), (17, 1), (18, 1), (19, 1), (20, 1), (21, 1), (22, 1), (23, 1), (24, 1), (25, 1), (26, 1), (27, 1), (28, 1), (29, 1), (30, 1), (31, 1), (32, 1), (33, 1)]]


# 7. Building the topic model

### Environment GCP 16vCPUs; 60GB RAM
### Training time 10798s(i.e. 3h) 

In [59]:
# Build LDA model
start = time.time()
lda_model = gensim.models.ldamodel.LdaModel(corpus=corpus,
                                           id2word=id2word,
                                           num_topics=20, 
                                           random_state=100,
                                           update_every=1,
                                           chunksize=100,
                                           passes=10,
                                           alpha='auto',
                                           per_word_topics=True)

end = time.time()
print ('Training Time using the corpus without lemmalization')
print(end - start)

Training Time using the corpus without lemmalization
10798.353863716125


# 8. View the topics in LDA model

In [113]:
# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())
doc_lda = lda_model[corpus]

[(0,
  '0.261*"exclusive" + 0.171*"cap" + 0.072*"hilarious" + 0.046*"yell" + '
  '0.036*"buddy" + 0.034*"huffingtonpost" + 0.023*"realistically" + '
  '0.018*"please" + 0.016*"optimistic" + 0.012*"tyrant"'),
 (1,
  '0.049*"woman" + 0.045*"life" + 0.031*"kill" + 0.025*"war" + 0.024*"human" + '
  '0.024*"die" + 0.023*"right" + 0.022*"abortion" + 0.015*"support" + '
  '0.015*"death"'),
 (2,
  '0.683*"reddit" + 0.237*"comment" + 0.059*"subreddit" + 0.003*"pm" + '
  '0.000*"submission" + 0.000*"question_regarde" + 0.000*"moderator" + '
  '0.000*"redd" + 0.000*"explicitly" + 0.000*"contribution"'),
 (3,
  '0.516*"speech" + 0.044*"flood" + 0.038*"downplay" + 0.036*"incompetence" + '
  '0.027*"seizure" + 0.026*"partial" + 0.025*"calculation" + 0.021*"assemble" '
  '+ 0.020*"falsehood" + 0.017*"redundant"'),
 (4,
  '0.061*"say" + 0.053*"article" + 0.041*"topic" + 0.031*"claim" + '
  '0.030*"fact" + 0.030*"news" + 0.026*"post" + 0.025*"comment" + '
  '0.024*"source" + 0.024*"story"'),
 (5,
  '0.

### Save model

In [82]:
#from gensim.test.utils import datapath
#temp_file = datapath("C:/Users/79998/Documents/")
#lda_model.save(temp_file)

import dill
with open("lda_model_2019_02.state",'wb') as f:
    dill.dump(lda_model,f)

### Load model

In [1]:
from gensim import corpora, models, similarities
model_foo = models.LdaModel.load("lda_model_2019_02.state")
from pprint import pprint
pprint(model_foo.print_topics())

ModuleNotFoundError: No module named 'dill'

# 9. Compute Model Perplexity and Coherence Score

In [89]:
# Compute Perplexity
print('\nPerplexity: ', lda_model.log_perplexity(corpus))  # a measure of how good the model is. lower the better.

# Compute Coherence Score
coherence_model_lda = CoherenceModel(model=lda_model, texts=data_lemmatized, dictionary=id2word, coherence='c_v')
coherence_lda = coherence_model_lda.get_coherence()
print('\nCoherence Score: ', coherence_lda)


Perplexity:  -10.474324421768227

Coherence Score:  0.42084377073517026


# 10. Visualization

In [90]:
# Visualize the topics
pyLDAvis.enable_notebook()
vis = pyLDAvis.gensim.prepare(lda_model, corpus, id2word)
vis

/usr/local/lib/python3.5/dist-packages/pyLDAvis/_prepare.py:257: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  return pd.concat([default_term_info] + list(topic_dfs))


PreparedData(topic_coordinates=            Freq  cluster  topics         x         y
topic                                                
9      47.634239        1       1  0.221393  0.088568
15      9.876889        1       2  0.178325  0.180785
16      7.762465        1       3  0.142741 -0.171661
13      6.013822        1       4  0.184270 -0.087275
8       5.646519        1       5  0.165419  0.042176
12      4.153207        1       6  0.048550 -0.206811
4       4.014362        1       7  0.059894 -0.170114
19      3.614823        1       8  0.145559 -0.064661
6       2.989926        1       9  0.118846  0.213293
1       2.575282        1      10  0.053842  0.153323
5       1.462642        1      11  0.008317 -0.218979
17      0.781696        1      12 -0.051361  0.077416
10      0.779545        1      13  0.027550  0.189069
18      0.722174        1      14 -0.159782 -0.016020
7       0.503510        1      15 -0.157507  0.000514
2       0.500309        1      16 -0.095115 -0.099002
11      0.460824        1      17 -0.096422 -0.057205
0       0.239655        1      18 -0.240408  0.044686
3       0.217691        1      19 -0.270714  0.047070
14      0.050420        1      20 -0.283395  0.054829, topic_info=      Category           Freq              Term          Total  loglift  \
16328  Default  301423.000000                be  301423.000000  30.0000   
49231  Default  184031.000000            remove  184031.000000  29.0000   
81786  Default  249025.000000           removal  249025.000000  28.0000   
81455  Default  241845.000000                 s  241845.000000  27.0000   
30066  Default  278722.000000               say  278722.000000  26.0000   
...        ...            ...               ...            ...      ...   
47871  Topic20       0.046819        explicitly       0.963569   4.5682   
30010  Topic20       0.046830        submission       0.978266   4.5533   
11061  Topic20       0.046824  question_regarde       0.970467   4.5611   
36816  Topic20       0.046821         moderator       0.967570   4.5641   
21944  Topic20       0.046818         recommend       0.963191   4.5686   

       logprob  
16328  30.0000  
49231  29.0000  
81786  28.0000  
81455  27.0000  
30066  26.0000  
...        ...  
47871 -12.6123  
30010 -12.6121  
11061 -12.6122  
36816 -12.6123  
21944 -12.6123  

[1021 rows x 6 columns], token_table=       Topic      Freq       Term
term                             
84052     15  0.998191      abide
9060      10  0.999927   abortion
50910     11  0.918841  abundance
74374      1  0.090421      abuse
74374      8  0.049861      abuse
...      ...       ...        ...
83478      9  0.995882   yearbook
61482     18  0.999750       yell
54335      6  0.999309      yield
59199      6  0.743066    youtube
59199     11  0.256609    youtube

[1352 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'ylab': 'PC2', 'xlab': 'PC1'}, topic_order=[10, 16, 17, 14, 9, 13, 5, 20, 7, 2, 6, 18, 11, 19, 8, 3, 12, 1, 4, 15])

# 11. Find the optimal number of topics for LDA

# 12. Find the dominant topic for each sentece

In [91]:
def format_topics_sentences(ldamodel=lda_model, corpus=corpus, texts=data):
    # Init output
    sent_topics_df = pd.DataFrame()

    # Get main topic in each document
    for i, row in enumerate(ldamodel[corpus]):
        row = sorted(row, key=lambda x: (x[1]), reverse=True)
        # Get the Dominant topic, Perc Contribution and Keywords for each document
        for j, (topic_num, prop_topic) in enumerate(row):
            if j == 0:  # => dominant topic
                wp = ldamodel.show_topic(topic_num)
                topic_keywords = ", ".join([word for word, prop in wp])
                sent_topics_df = sent_topics_df.append(pd.Series([int(topic_num), round(prop_topic,4), topic_keywords]), ignore_index=True)
            else:
                break
    sent_topics_df.columns = ['Dominant_Topic', 'Perc_Contribution', 'Topic_Keywords']

    # Add original text to the end of the output
    contents = pd.Series(texts)
    sent_topics_df = pd.concat([sent_topics_df, contents], axis=1)
    return(sent_topics_df)

In [117]:
df_topic_sents_keywords = format_topics_sentences(ldamodel=lda_model, corpus=corpus, texts=data)

# Format
df_dominant_topic = df_topic_sents_keywords.reset_index()
df_dominant_topic.columns = ['Document_No', 'Dominant_Topic', 'Topic_Perc_Contrib', 'Keywords', 'Text']

# Show
df_dominant_topic.head(10)

TypeError: unorderable types: int() < tuple()

In [99]:
for i,row in enumerate(lda_model[corpus[:1]]):
    print(row)

([(1, 0.039706282), (2, 0.011873434), (4, 0.024813917), (6, 0.039462317), (8, 0.043480113), (9, 0.43017164), (11, 0.015641183), (12, 0.026787803), (13, 0.05003716), (15, 0.17303813), (16, 0.092475235), (17, 0.019927787), (19, 0.018837687)], [(0, [1, 15]), (1, [9, 8]), (2, [1, 12]), (3, [15, 16, 19]), (4, [17, 15, 9]), (5, [9, 16]), (6, [2, 4]), (7, [16, 8]), (8, [16, 8]), (9, [9, 15]), (10, [16]), (11, [15]), (12, [9]), (13, [15, 6, 9, 13]), (14, [9, 15, 8]), (15, [9, 15]), (16, [13, 6]), (17, [15]), (18, [15, 9]), (19, []), (20, [9, 15]), (21, [11]), (22, [13]), (23, [17, 15, 6, 8]), (24, [9, 16, 13]), (25, [9, 15]), (26, [15, 6, 9]), (27, []), (28, [15]), (29, [9, 15, 1, 8, 13]), (30, [9]), (31, [8, 6]), (32, [15]), (33, [9, 15]), (34, [9])], [(0, [(1, 0.8036822), (15, 0.19518264)]), (1, [(8, 0.037463937), (9, 0.9623667)]), (2, [(1, 0.77516407), (12, 0.22134107)]), (3, [(15, 0.7164559), (16, 0.23787117), (19, 0.04428897)]), (4, [(9, 0.18026282), (15, 0.24947219), (17, 0.570165)]), (5

In [112]:
infer_topics = lda_model[corpus[:2]]
print(infer_topics)

In [109]:
for infer_topic in infer_topics:
    print("{}\t{}\n".format(lda_model.print_topic(infer_topic[0]), infer_topic[1]))

/usr/local/lib/python3.5/dist-packages/gensim/models/ldamodel.py:1224: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  topic = self.get_topics()[topicid]


IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [125]:
corpus[:5]

[[(0, 1),
  (1, 1),
  (2, 1),
  (3, 1),
  (4, 1),
  (5, 1),
  (6, 1),
  (7, 1),
  (8, 1),
  (9, 1),
  (10, 2),
  (11, 1),
  (12, 1),
  (13, 1),
  (14, 1),
  (15, 1),
  (16, 1),
  (17, 1),
  (18, 1),
  (19, 1),
  (20, 1),
  (21, 1),
  (22, 1),
  (23, 1),
  (24, 1),
  (25, 1),
  (26, 1),
  (27, 1),
  (28, 1),
  (29, 1),
  (30, 1),
  (31, 1),
  (32, 1),
  (33, 1),
  (34, 1)],
 [(18, 1),
  (20, 2),
  (24, 3),
  (33, 1),
  (35, 1),
  (36, 1),
  (37, 2),
  (38, 1),
  (39, 1),
  (40, 1),
  (41, 1),
  (42, 1),
  (43, 1),
  (44, 1),
  (45, 1),
  (46, 1),
  (47, 1),
  (48, 1),
  (49, 1),
  (50, 1),
  (51, 1),
  (52, 1),
  (53, 2),
  (54, 2),
  (55, 1),
  (56, 1),
  (57, 1),
  (58, 1),
  (59, 1),
  (60, 2),
  (61, 1),
  (62, 1),
  (63, 1),
  (64, 2),
  (65, 1),
  (66, 2),
  (67, 1),
  (68, 1),
  (69, 1),
  (70, 1),
  (71, 1),
  (72, 1),
  (73, 1),
  (74, 1),
  (75, 2),
  (76, 1),
  (77, 2),
  (78, 1),
  (79, 2),
  (80, 1),
  (81, 3),
  (82, 1),
  (83, 2),
  (84, 1),
  (85, 1),
  (86, 2),
  (87, 1

In [148]:
count_foo = 0
for doc in corpus[:5]:
    chunk = []
    chunk.append(doc)
    print(lda_model.inference(chunk, collect_sstats=False))

(array([[ 0.086895  ,  2.8373368 ,  0.8477928 ,  0.07149873,  1.7738649 ,
         0.25506178,  2.8196368 ,  0.27326912,  3.107274  , 30.73978   ,
         0.1819211 ,  1.1177117 ,  1.9142535 ,  3.5757318 ,  0.03731016,
        12.3659725 ,  6.608171  ,  1.423502  ,  0.07641966,  1.3461279 ]],
      dtype=float32), None)
(array([[8.6895004e-02, 1.9915497e+00, 9.2316166e-02, 1.0714285e+00,
        1.7256030e+00, 2.5506178e-01, 5.6854649e+00, 1.2709453e+00,
        1.3924249e+01, 1.0878944e+02, 1.8203606e-01, 1.3515106e-01,
        3.7718568e+00, 5.2204561e+00, 3.7310164e-02, 1.7791197e+01,
        2.1556547e+01, 1.6831012e+00, 1.0679431e+00, 4.0598893e+00]],
      dtype=float32), None)
(array([[ 7.0861263 ,  5.5801897 ,  0.09231617,  0.07149873,  2.53102   ,
         0.25506178,  9.155501  ,  0.27326912,  2.9051547 , 34.998817  ,
         0.18249866,  0.13515106,  1.7006612 ,  3.073293  ,  0.03731016,
         3.3538744 ,  5.2265835 ,  0.32793778,  0.07641966,  1.4443938 ]],
      dtype

In [135]:
lda_model.inference(corpus[1:2], collect_sstats=False)

(array([[8.6895004e-02, 1.9915504e+00, 9.2316166e-02, 1.0714285e+00,
         1.7256029e+00, 2.5506178e-01, 5.6854744e+00, 1.2709453e+00,
         1.3924260e+01, 1.0878947e+02, 1.8203606e-01, 1.3515106e-01,
         3.7718575e+00, 5.2204542e+00, 3.7310164e-02, 1.7791189e+01,
         2.1556553e+01, 1.6831014e+00, 1.0679431e+00, 4.0598292e+00]],
       dtype=float32), None)

In [146]:
chunk_foo = []
chunk_foo.append(corpus[1])

In [150]:
lda_model.get_document_topics(corpus[0])

[(1, 0.039706856),
 (2, 0.011870805),
 (4, 0.024816532),
 (6, 0.039440483),
 (8, 0.043486964),
 (9, 0.43017),
 (11, 0.015641183),
 (12, 0.026787607),
 (13, 0.050042246),
 (15, 0.17303197),
 (16, 0.09247361),
 (17, 0.019946678),
 (19, 0.018837733)]

In [151]:
lda_model.get_document_topics(corpus[1])

[(1, 0.010459725),
 (6, 0.029861595),
 (8, 0.07313348),
 (9, 0.5713812),
 (12, 0.019810453),
 (13, 0.027418688),
 (15, 0.09344158),
 (16, 0.11321884),
 (19, 0.02131712)]

In [159]:
for i in range(len(corpus[:5])):
    print(lda_model.get_document_topics(corpus[i]))

[(1, 0.039706387), (2, 0.01187258), (4, 0.024814766), (6, 0.039440557), (8, 0.043487206), (9, 0.43016967), (11, 0.015641183), (12, 0.026787763), (13, 0.05004227), (15, 0.17303464), (16, 0.09247346), (17, 0.019944502), (19, 0.018837715)]
[(1, 0.010459524), (6, 0.029862612), (8, 0.07313534), (9, 0.57138604), (12, 0.019810626), (13, 0.027418815), (15, 0.093441084), (16, 0.1132199), (19, 0.021308778)]
[(0, 0.09026099), (1, 0.071085915), (4, 0.032239385), (6, 0.1166286), (8, 0.036982656), (9, 0.44580284), (12, 0.021662522), (13, 0.039146397), (15, 0.042721245), (16, 0.066582814), (19, 0.018398324)]
[(1, 0.026415836), (4, 0.039751884), (6, 0.029127419), (8, 0.04862143), (9, 0.5418107), (12, 0.04395055), (13, 0.05162396), (15, 0.07755023), (16, 0.06742834), (19, 0.033797123)]
[(1, 0.040498942), (4, 0.02804048), (5, 0.01212842), (6, 0.018488329), (8, 0.07134947), (9, 0.50354177), (12, 0.028106129), (13, 0.059478205), (15, 0.072182044), (16, 0.12976684), (19, 0.020812301)]


In [161]:
sent_topics_df = pd.DataFrame()
for i in range(len(corpus[:5])):
    #print(lda_model.get_document_topics(corpus[i]))
    for (topic_num, prop_topic) in enumerate(row):
                #if j == 0:  # => dominant topic
                wp = lda_model.show_topic(topic_num)
                topic_keywords = ", ".join([word for word, prop in wp])
                sent_topics_df = sent_topics_df.append(pd.Series([int(topic_num), round(prop_topic,4), topic_keywords]), ignore_index=True)

TypeError: type list doesn't define __round__ method

In [ ]:
topic_figure_df = pd.DataFrame()
topic_figure_df.columns = ['Author_Num']
for i in range(20):
    topic_figure_df.colunms.append("{}".format(i+1))
topic_figure_df.columns = ['Author_Num']
topic_figure_df = pd.concat(topic_figure_df, contents)
    # Add original text to the end of the output
    contents = pd.Series(texts)
    sent_topics_df = pd.concat([sent_topics_df, contents], axis=1)
    return(sent_topics_df)


In [190]:
topic_figure_df = pd.DataFrame(index=np.arange(len(corpus)), columns=np.arange(21))
columns_title = ['Author_Num']
for i in range(20):
    columns_title.append("{}".format(i+1))
columns_title
topic_figure_df.columns = columns_title


In [191]:
topic_figure_df.head()


Author_Num    1    2    3    4    5    6    7    8    9  ...   11   12   13  \
0        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
2        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
3        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

    14   15   16   17   18   19   20  
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 21 columns]

In [199]:
corpus_len = len(corpus)
start = time.time()
for i in range(corpus_len):
    topic_figure_df.iloc[i,0] = i
    #row = lda_model.get_document_topics(corpus[i])
    for j,(topic_num, prop_topic) in enumerate(row):
        topic_figure_df.iloc[i,topic_num] = prop_topic
    print("{}/{}".format(i,corpus_len))
end = time.time()
print ('Time to find the topic for each users comments')
print(end - start)

0/199482
1/199482
2/199482
3/199482
4/199482
5/199482
6/199482
7/199482
8/199482
9/199482
10/199482
11/199482
12/199482


KeyboardInterrupt: 

Author_Num          1          2    3          4    5          6    7  \
0          0  0.0397067  0.0118728  NaN  0.0248146  NaN  0.0394496  NaN   
1          1  0.0104599        NaN  NaN        NaN  NaN   0.029861  NaN   
2        NaN        NaN        NaN  NaN        NaN  NaN        NaN  NaN   
3        NaN        NaN        NaN  NaN        NaN  NaN        NaN  NaN   
4        NaN        NaN        NaN  NaN        NaN  NaN        NaN  NaN   

           8         9  ...         11         12         13   14         15  \
0  0.0434838  0.430171  ...  0.0156412  0.0267877    0.05004  NaN   0.173031   
1  0.0731324  0.571378  ...        NaN  0.0198104  0.0274186  NaN  0.0934419   
2        NaN       NaN  ...        NaN        NaN        NaN  NaN        NaN   
3        NaN       NaN  ...        NaN        NaN        NaN  NaN        NaN   
4        NaN       NaN  ...        NaN        NaN        NaN  NaN        NaN   

          16         17   18         19   20  
0  0.0924744  0.0199422  NaN  0.0188377  NaN  
1   0.113218        NaN  NaN   0.021322  NaN  
2        NaN        NaN  NaN        NaN  NaN  
3        NaN        NaN  NaN        NaN  NaN  
4        NaN        NaN  NaN        NaN  NaN  

[5 rows x 21 columns]

In [216]:
import magic
corpus_len = len(corpus[:3])
out_file = "comments_topic_2019_02"
start = time.time()
with open(out_file, 'w') as outfile:
    for i in range(corpus_len):
        row = "{}".format(i)+"/t"+ str(lda_model.get_document_topics(corpus[i]))
        print("{}/{}".format(i,corpus_len))
    json.dump(row, outfile)      
end = time.time()
print ('Time to find the topic for each users comments')
print(end - start)

0/3
1/3
2/3
Time to find the topic for each users comments
0.004186868667602539
